File: leetcode/101-200/152-maximum-product-subarray/max_product.py
"""LeetCode 152 — Maximum Product Subarray (Python mirror / oracle).

O(n) DP tracking both the max and min running product ending at each index (a
negative element swaps their roles). Mirrors the Kāra version exactly.
"""


def max_product(nums):
    n = len(nums)
    if n == 0:
        return 0
    best = cur_max = cur_min = nums[0]
    for i in range(1, n):
        x = nums[i]
        if x < 0:
            cur_max, cur_min = cur_min, cur_max
        cur_max = max(x, cur_max * x)
        cur_min = min(x, cur_min * x)
        if cur_max > best:
            best = cur_max
    return best


def run(nums):
    print(max_product(nums))


def main():
    run([2, 3, -2, 4])
    run([-2, 0, -1])
    run([-2, 3, -4])
    run([2, -5, -2, -4, 3])
    run([-2])
    run([0, 2])


main()

File: leetcode/101-200/152-maximum-product-subarray/bench/max_product.rs
//! Benchmark workload — Maximum Product Subarray, O(n) running max/min DP.
//! Algorithmic mirror of bench/max_product.kara / .c / .py / go-seq. See
//! ../README.md § Benchmarks for N / K and the deterministic LCG generator.

fn max_product(nums: &[i64]) -> i64 {
    if nums.is_empty() {
        return 0;
    }
    let mut best = nums[0];
    let mut cur_max = nums[0];
    let mut cur_min = nums[0];
    for &x in &nums[1..] {
        if x < 0 {
            std::mem::swap(&mut cur_max, &mut cur_min);
        }
        cur_max = x.max(cur_max * x);
        cur_min = x.min(cur_min * x);
        if cur_max > best {
            best = cur_max;
        }
    }
    best
}

fn main() {
    const N: usize = 2_000_000;
    let mut data = vec![0i64; N];
    let mut state: i64 = 12345;
    for d in data.iter_mut() {
        state = (state.wrapping_mul(1103515245).wrapping_add(12345)) & 2_147_483_647;
        *d = (state % 5) - 2;
    }
    let mut sum: i64 = 0;
    for _ in 0..10 {
        sum += max_product(&data);
    }
    println!("{}", sum);
}
